libs/wire-api/src/Wire/API/Error.hs

```diff
@@ -0,0 +1,287 @@
+  where
+    err = dynError' e
+    label = eLabel err
+    code = eCode err
+    message = eMessage err
+
+    addExample = S.schema . S.example ?~ A.toJSON e
+    labelSchema :: ValueSchema SwaggerDoc Text
+    labelSchema = unnamed $ enum @Text "Label" (element label label)
+    codeSchema :: ValueSchema SwaggerDoc Natural
+    codeSchema = unnamed $ enum @Natural "Status" (element code code)
+
+instance KnownError e => ToSchema (SStaticError e) where
+  schema = staticErrorSchema seSing
+
+data CanThrow e
+
+data CanThrowMany e
+
+instance RoutesToPaths api => RoutesToPaths (CanThrow err :> api) where
+  getRoutes = getRoutes @api
+
+instance RoutesToPaths api => RoutesToPaths (CanThrowMany errs :> api) where
+  getRoutes = getRoutes @api
+
+instance (HasServer api ctx) => HasServer (CanThrow e :> api) ctx where
+  type ServerT (CanThrow e :> api) m = ServerT api m
+
+  route _ = route (Proxy @api)
+  hoistServerWithContext _ = hoistServerWithContext (Proxy @api)
+
+instance (HasServer api ctx) => HasServer (CanThrowMany es :> api) ctx where
+  type ServerT (CanThrowMany es :> api) m = ServerT api m
+
+  route _ = route (Proxy @api)
+  hoistServerWithContext _ = hoistServerWithContext (Proxy @api)
+
+instance
+  (HasSwagger api, IsSwaggerError e) =>
+  HasSwagger (CanThrow e :> api)
+  where
+  toSwagger _ = addToSwagger @e (toSwagger (Proxy @api))
+
+instance HasSwagger api => HasSwagger (CanThrowMany '() :> api) where
+  toSwagger _ = toSwagger (Proxy @api)
+
+instance
+  (HasSwagger (CanThrowMany es :> api), IsSwaggerError e) =>
+  HasSwagger (CanThrowMany '(e, es) :> api)
+  where
+  toSwagger _ = addToSwagger @e (toSwagger (Proxy @(CanThrowMany es :> api)))
+
+type family DeclaredErrorEffects api :: EffectRow where
+  DeclaredErrorEffects (CanThrow e :> api) = (ErrorEffect e ': DeclaredErrorEffects api)
+  DeclaredErrorEffects (CanThrowMany '(e, es) :> api) =
+    DeclaredErrorEffects (CanThrow e :> CanThrowMany es :> api)
+  DeclaredErrorEffects (x :> api) = DeclaredErrorEffects api
+  DeclaredErrorEffects (Named n api) = DeclaredErrorEffects api
+  DeclaredErrorEffects api = '[]
+
+errorResponseSwagger :: forall e. KnownError e => S.Response
+errorResponseSwagger =
+  mempty
+    & S.description .~ (eMessage err <> " (label: `" <> eLabel err <> "`)")
+    & S.schema ?~ S.Inline (S.toSchema (Proxy @(SStaticError e)))
+  where
+    err = dynError @e
+
+addStaticErrorToSwagger :: forall e. KnownError e => S.Swagger -> S.Swagger
+addStaticErrorToSwagger =
+  S.allOperations . S.responses . S.responses . at (fromIntegral (eCode err))
+    %~ Just . addRef
+  where
+    err = dynError @e
+    resp = errorResponseSwagger @e
+
+    addRef :: Maybe (S.Referenced S.Response) -> S.Referenced S.Response
+    addRef Nothing = S.Inline resp
+    addRef (Just (S.Inline resp1)) = S.Inline (combineResponseSwagger resp1 resp)
+    addRef (Just r@(S.Ref _)) = r
+
+type family MapError (e :: k) :: StaticError
```

Comment:
I don't get this one. What is it for? Can you clarify?

Reply:
This is to convert a custom error kind to a static error. See for example `Wire.API.Error.Galley` (introduced by the second commit).